core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -440,6 +448,7 @@ public void run() {
 
   @Override
   public final void cancel(Status reason) {
+    cancellationStatus = reason;
```

Comment:
Move this to just before the synchronized block (or as an else within the synchronized block)? Right now if commit() fails the status is still set, which may be confusing.